File: DS/Stacks/cpp/infix_to_prefix.cpp
#include <bits/stdc++.h>
using namespace std;

int prec(char c){
    if(c == '*' || c == '/') 
    return 1;

    else if(c == '+' || c == '-') 
    return 0;

    else
    return -1;
}

string infixToPrefix(string s){
    stack<char> st;
    int i;

    string str;

    for (i = s.length()-1; i >=0; i--)
    {
        if(iswalnum(s[i])){
           str += s[i]; 
        }

        else if(s[i]==')'){
            st.push(s[i]);
        }

        else if(s[i]=='('){
            while (st.top()!=')')
            {
                str += st.top();
                st.pop();
            }
            st.pop();
        }

        else {
            while(!st.empty() && prec(s[i]) <= prec(st.top())) {
                str += st.top();
                st.pop(); 
            }
            st.push(s[i]);
        }
    }

    while (!st.empty())
    {
        str += st.top();
        st.pop();
    }

    reverse(str.begin(),str.end());
    return str;
}

int main(){
    string s = "((A*B)+(C*D))-E";
    cout<<infixToPrefix(s);
}

// output : -+*AB*CDE

File: DS/Stacks/Python/infix_to_prefix.py
def pre(ch):
    if ch == '*' or ch == '/':
        return 1
    elif ch == '+' or ch == '-':
        return 0
    else:
        return -1


def infix_to_prefix(s, n):
    stack = []
    top = -1
    n -= 1

    result = ""
    while n >= 0:
        if s[n].isalnum():
            result += s[n]

        elif s[n] == ')':
            stack.append(s[n])
            top += 1

        elif s[n] == '(':
            while stack[top] != ')':
                result += stack[top]
                top -= 1
                stack.pop()
            top -= 1
            stack.pop()

        else:
            while len(stack) != 0 and pre(s[n]) <= pre(stack[top]):
                result += stack[top]
                stack.pop()
                top -= 1
            stack.append(s[n])
            top += 1
        n -= 1

    while len(stack) != 0:
        result += stack.pop()

    return result[::-1]


s = "((A*B)+(C*D))-E"
print(infix_to_prefix(s, len(s)))

# output: -+*AB*CDE